### ISSUE TITLE
Textarea inputfield with view-only access won't enforce any height (rows) limit

### ISSUE THREAD

**Issue by er314** [processwire/processwire-issues#1627]
### Short description of the issue

When a backend user has view-only persmission on a Textarea inputfield, 
and setting _"show field in page editor if viewable but not editable"_ is enabled,
the output of the field contents in the page editor will take whatever vertical space needed for its contents, thereby ignoring the "Rows" setting for this field.
For example, if the textarea field is set to 5 rows, and its contents is 200 lines long, then :
	. -OK- its "read-write" rendering will be a textarea html element with 5 rows, and a vertical scrollbar for scrolling down, until the 200th line.
	. -NOK- its "read-only" rendering will be straight raw text of 200 lines -> UX-wise, it's a mess.

### Expected behavior

The "read-only" rendering should somehow respect the "Rows" setting, so that the general layout of the page edit screen is preserved.

### Optional: Suggestion for a possible fix

I guess we have at least 2 possible methods for fixing the "read-only" rendering :  
. it could consist in wrapping the text inside a textarea element, with attributes "readonly" and "rows" properly set.
. or it could consist in wrapping the text inside a div element, and applying a "height constraint" (height = nbRows * line-height em) + scrollbar (overflow:scroll) on it.

Note : All the above applies to textarea field when configured with type = "Textarea". It's also basically the same problem when its type is "CKEditor".

### Setup/Environment

Processwire 3.0.203, default admin theme

**Comment by teppokoivula** [comment]
This is a case of "it depends".

I've relied many times on this behaviour, i.e. assumed that it's by design that if textarea is set non-editable, it should be displayed as "inline text". If those textareas were to change now, that would result in (at least minor visual) glitches on existing sites.

Assuming that there's an easy way to limit the height in the first place, perhaps it needs to be a field-specific option?

**Comment by er314** [comment]
Yes, I understand that the current rendering may be hooked upon, therefore any changes to it may have an impact.
Notwithstanding the impact on existing hooks, do you have some use cases where the current rendering, with unlimited height, is *making sense* ?
I see none ; I see this unlimited field height as a messy rendering of the page edit layout.

By the way, I can fix it live by hooks (that's what I'm doing actually).
So the question is about what makes sense as standard rendering (and how to avoid impact on legacy code).

**Comment by er314** [comment]
> Assuming that there's an easy way to limit the height in the first place, perhaps it needs to be a field-specific option?

If no better idea, in the "Rows" configuration of the field, addition of a setting checkbox "Also apply to VIEW mode ?", unchecked by default 
-> this ensures non regression

**Comment by BernhardBaumrock** [comment]
I agree with teppo :) use case: Superuser can input content in a textarea and the user can only read this text (for example some kind of backend docs / help pages)

**Comment by er314** [comment]
Ok, you have a use case where it makes sense to have unlimited height for a read-only textarea.
I guess this is only relevant when field is configured at 100% width, otherwise again it would end up as a messed-up page layout.
So do you see an interest in having the capability to choose between unlimited height OR "Rows" height ?
Or am I the only one thinking this way ?

**Comment by szabeszg** [comment]
> Or am I the only one thinking this way ?

No, you are not the only one. I also think that if something is set to be limited, then it should be limited the way it was set to be.
As for relying on inconsistent behavior can cause troubles later on when things are streamlined to be consistent is definitely an issue but should prevent us from trying to streamline things.

> Superuser can input content in a textarea and the user can only read this text (for example some kind of backend docs / help pages)

Superuser could explicitly set the hight to maximum (there could be such an option added). Superuser is not encouraged to rely on half baked features being set in stone, rather, superuser could at least open a feature request over here ;)

**Comment by teppokoivula** [comment]
> As for relying on inconsistent behavior can cause troubles later on when things are streamlined to be consistent is definitely an issue but should prevent us from trying to streamline things.

> Superuser could explicitly set the hight to maximum (there could be such an option added). Superuser is not encouraged to rely on half baked features being set in stone, rather, superuser could at least open a feature request over here ;)

Again, I would argue that this is debatable. I've always considered row limit to be applicable to the input, as in <textarea> (or whatever it is that CKEditor uses for display). As such, I don't consider this half-baked at all, but rather intentional behaviour. Non-editable fields are intentionally as simple/plain as possible (and often just plain text), while limiting the height would at the very least need to rely on CSS styles.

I can see how this could be seen the other way as well, of course :)

The fact remains that changing this behaviour now would be very likely to break existing implementations. As such — and considering that some of my own sites would suffer from this as well — I would be against changing it without some sort of opt-in method.

**Comment by er314** [comment]
Therefore against all odds we should easily agree on this universal & backwards-compatible approach : 
Having the capability to choose between unlimited height OR "Rows" height.
-> in the "Rows" configuration of the field, addition of a setting checkbox "Also apply to VIEW mode ?", unchecked by default.
Anyone disagree ? :relaxed:

**Comment by er314** [comment]
As current period is welcoming the Re-Up'ing of pending issues (see latest weekly update – 14 July 2023),
I Re-Up this one.

Here  is a screenshot of the issue, when user has view-only permission on a textarea field
-> the global page layout is broken due to this field :  

![textarea-overflow-when-ReadOnly](https://github.com/processwire/processwire-issues/assets/8220747/f9367837-f560-43d8-9afb-4eb2a12ff723)

Again, we can live with it, as it's fixable by hook.
Yet, the valid way forward should be that the Core ensures correct layout rendering, for view-only textarea fields.

As the current behavior is in active use by some projects, the fix would probably consist in something along the lines of this universal & backwards-compatible approach :
Having the capability to choose between unlimited height OR "Rows" height.
-> in the "Rows" configuration of the field, addition of a setting checkbox "Also apply to VIEW mode ?", unchecked by default.

**Comment by matjazpotocnik** [comment]
@ryancramerdesign 

Implemented the backward-compatible opt-in solution.

The new **Limit height in view mode** checkbox is unchecked by default. When enabled, view-only textarea output uses the configured `Rows` value as its maximum height and adds scrolling only when content overflows. The constraint is applied after `renderValue()`, so it works consistently for plain textarea, CKEditor, and TinyMCE while preserving their existing output markup.

Changed:

- `InputfieldTextarea.module:12`: setting, configuration checkbox, height wrapper helper, context support, and version bump to 104.
- `InputfieldWrapper.php:1282`: applies the constraint to all textarea-derived inputfields.
- `InputfieldTextarea.test.php:130`: regression coverage for compatibility, configuration, plain text, CKEditor, and TinyMCE.
- `API.md:38`: documented the new setting.

**Suggested fix**
```diff
diff --git a/wire/core/Inputfield/InputfieldWrapper.php b/wire/core/Inputfield/InputfieldWrapper.php
--- a/wire/core/Inputfield/InputfieldWrapper.php
+++ b/wire/core/Inputfield/InputfieldWrapper.php
@@
 		// renderValueMode
 		$out = $inputfield->renderValue();
+		if($inputfield instanceof InputfieldTextarea) $out = $inputfield->applyRenderValueRows($out);
 		if(!$this->getSetting('renderValueMode')) $out .= $this->renderInputfieldDependencyValue($inputfield);

diff --git a/wire/modules/Inputfield/InputfieldTextarea/InputfieldTextarea.module b/wire/modules/Inputfield/InputfieldTextarea/InputfieldTextarea.module
--- a/wire/modules/Inputfield/InputfieldTextarea/InputfieldTextarea.module
+++ b/wire/modules/Inputfield/InputfieldTextarea/InputfieldTextarea.module
@@
  * @property int $rows Number of rows for textarea (default=5)
+ * @property bool $renderValueRows Limit view-mode output height to the configured rows? (default=false)
  * @property int $contentType Content type, applicable when used with FieldtypeTextarea. See FieldtypeTextarea contentType constants (default=0)
@@
-			'version' => 103,
+			'version' => 104,
@@
 		parent::init();
 		$this->setAttribute('rows', self::defaultRows);
+		$this->set('renderValueRows', false);
 		$this->setAttribute('maxlength', $this->getDefaultMaxlength());
@@
 		return $out;
 	}
+
+	/**
+	 * Apply configured rows as maximum height to view-mode markup
+	 *
+	 * @param string $out
+	 * @return string
+	 *
+	 */
+	public function applyRenderValueRows($out) {
+		if(!$this->renderValueRows || !strlen($out)) return $out;
+		$height = max(1, (int) $this->attr('rows')) * 2;
+		return "<div class='InputfieldTextareaRenderValueRows' style='max-height:{$height}em;overflow:auto'>$out</div>";
+	}
@@
 		if($field->attr('value') == self::defaultRows) $field->collapsed = Inputfield::collapsedYes;
 		$inputfields->append($field);
+
+		/** @var InputfieldCheckbox $field */
+		$field = $this->wire()->modules->get('InputfieldCheckbox');
+		$field->setAttribute('name', 'renderValueRows');
+		$field->label = $this->_('Limit height in view mode');
+		$field->description = $this->_('Use the Rows setting as the maximum height when this field is view-only.');
+		if($this->renderValueRows) $field->setAttribute('checked', 'checked');
+		$inputfields->append($field);
@@
-		return array_merge(parent::___getConfigAllowContext($field), array('rows'));
+		return array_merge(parent::___getConfigAllowContext($field), array('rows', 'renderValueRows'));
 	}

diff --git a/wire/modules/Inputfield/InputfieldTextarea/InputfieldTextarea.test.php b/wire/modules/Inputfield/InputfieldTextarea/InputfieldTextarea.test.php
--- a/wire/modules/Inputfield/InputfieldTextarea/InputfieldTextarea.test.php
+++ b/wire/modules/Inputfield/InputfieldTextarea/InputfieldTextarea.test.php
@@
 		$this->testRenderValue();
 		$this->testRenderValueHTML();
+		$this->testRenderValueRows();
+		$this->testRenderValueRowsConfig();
 		$this->testRowsAttribute();
@@
+	protected function testRenderValueRows() {
+		$f = $this->newInputfield();
+		$f->rows = 5;
+		$f->val("Line 1\nLine 2");
+		$wrapper = $this->wire(new InputfieldWrapper());
+		$wrapper->add($f);
+		$html = $wrapper->renderValue();
+		$this->check('view-mode rows limit is disabled by default', false, strpos($html, 'InputfieldTextareaRenderValueRows') !== false);
+
+		$f->renderValueRows = true;
+		$html = $wrapper->renderValue();
+		$this->check('view-mode rows limit adds overflow wrapper', 'InputfieldTextareaRenderValueRows', $html, '*=');
+		$this->check('view-mode rows limit uses configured rows', 'max-height:10em', $html, '*=');
+		$this->check('view-mode rows limit preserves rendered value', 'Line 1<br', $html, '*=');
+
+		foreach(array('InputfieldCKEditor', 'InputfieldTinyMCE') as $class) {
+			$f = $this->wire()->modules->get($class);
+			if(!$f) continue;
+			$f->rows = 5;
+			$f->renderValueRows = true;
+			$f->val('<p>Rich text</p>');
+			$wrapper = $this->wire(new InputfieldWrapper());
+			$wrapper->add($f);
+			$html = $wrapper->renderValue();
+			$this->check("view-mode rows limit applies to $class", 'InputfieldTextareaRenderValueRows', $html, '*=');
+		}
+	}
+
+	protected function testRenderValueRowsConfig() {
+		$f = $this->newInputfield();
+		$config = $f->getConfigInputfields();
+		$field = $config->getChildByName('renderValueRows');
+		$this->check('view-mode rows limit has config checkbox', true, $field instanceof InputfieldCheckbox);
+		$this->check('view-mode rows limit config is unchecked by default', false, (bool) $field->attr('checked'));
+		$this->check('view-mode rows limit supports field context', true, in_array('renderValueRows', $f->getConfigAllowContext(null)));
+	}
+
diff --git a/wire/modules/Inputfield/InputfieldTextarea/API.md b/wire/modules/Inputfield/InputfieldTextarea/API.md
--- a/wire/modules/Inputfield/InputfieldTextarea/API.md
+++ b/wire/modules/Inputfield/InputfieldTextarea/API.md
@@
 - Renders `<textarea>` instead of `<input>`
 - Supports `rows` property (default 5)
+- Can optionally limit view-mode output to the configured number of rows
@@
-| Property      | Type  | Default | Description                                                              |
-|---------------|-------|---------|--------------------------------------------------------------------------|
-| `rows`        | `int` | `5`     | Number of rows shown for the textarea                                    |
-| `contentType` | `int` | `0`     | Content type when used with FieldtypeTextarea (see constants below)      |
+| Property          | Type   | Default | Description                                                         |
+|-------------------|--------|---------|---------------------------------------------------------------------|
+| `rows`            | `int`  | `5`     | Number of rows shown for the textarea                               |
+| `renderValueRows` | `bool` | `false` | Limit view-mode output height to the configured number of rows      |
+| `contentType`     | `int`  | `0`     | Content type when used with FieldtypeTextarea (see constants below) |
@@
+When `renderValueRows` is enabled, view-mode output is constrained to the
+height represented by `rows` and gains a scrollbar only when needed. It is
+disabled by default to preserve existing view-mode output.
+
+```php
+$f->rows = 5;
+$f->renderValueRows = true;
+```
+
+The constraint is applied by `InputfieldWrapper`, so it also applies to
+textarea-derived rich text editors that provide their own `renderValue()`
+implementation, including TinyMCE and CKEditor.
@@
 - The `rows` attribute controls the visual height of the textarea. The browser adds a scrollbar when content exceeds the visible area.
+- The `renderValueRows` setting is opt-in and affects values rendered through an `InputfieldWrapper`.
```
